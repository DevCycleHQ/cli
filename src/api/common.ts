export const BASE_URL = 'https://api.devcycle.com/'
export const AUTH_URL = 'https://auth.devcycle.com/'

export const buildHeaders = (token: string) => ({
    'Content-Type': 'application/json',
    Authorization: token,
})
