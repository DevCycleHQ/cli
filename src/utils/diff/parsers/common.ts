import parse from 'parse-diff'
import { ParseOptions, VariableMatch } from './types'

export abstract class BaseParser {
    abstract identity: string
    clientNames: string[]
    abstract variableMethodPattern: RegExp
    abstract variableNameCapturePattern: RegExp

    constructor(protected options: ParseOptions) {
        this.clientNames = options.clientNames ?? ['dvcClient']
    }

    buildRegexPattern() {
        return new RegExp(
            new RegExp(`(?:${this.clientNames.join('|')})`).source
            + this.variableMethodPattern.source
            + this.variableNameCapturePattern.source
        )
    }

    formatMatch(name: string, file: parse.File, change: parse.Change): VariableMatch {
        return {
            name,
            fileName: file.to ?? '',
            line: change.type === 'normal' ? change.ln1 : change.ln,
            mode: change.type === 'add' ? 'add' : 'remove'
        }
    }

    abstract match(content: string, options: ParseOptions): string | null

    parse(file: parse.File, options: ParseOptions = {}): VariableMatch[] {
        const results: VariableMatch[] = []
        for (const chunk of file.chunks) {
            for (const change of chunk.changes) {
                const match = this.match(change.content, options)
                if (match) {
                    results.push(this.formatMatch(match, file, change))
                }
            }
        }
        return results
    }
}
