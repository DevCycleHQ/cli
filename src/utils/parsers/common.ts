import parse from 'parse-diff'
import { ParseOptions, VariableDiffMatch, Range, MultilineUsageChunk, VariableUsageMatch } from './types'
import * as usage from '../../commands/usages/types'

type MultilineDiffChunk = {
    content: string
    changes: ParsedChange[]
}

type MatchWithRange = {
    // Variable name
    name: string
    // Start/end indicies of matching substring
    range: Range
    // true if the found key is not wrapped in quotes
    isUnknown?: boolean
}

export type MatchResult = {
    // true if the found key is not wrapped in quotes
    isUnknown?: boolean
    // content matching the pattern
    content: string
    // variable key found within match
    name: string
    // index of matching substring within parent string
    index: number
}

class ParsedChange {
    type: string
    content: string
    ln: number
    range: Range = { start: 0, end: 0 }

    constructor(change: parse.Change) {
        this.type = change.type
        this.content = change.content
        this.ln = change.type === 'normal' ? change.ln2 : change.ln
    }

    parseContent() {
        return this.type === 'normal'
            ? this.content.trim()
            : this.content.slice(1).trim()
    }

    isComment(commentCharacters: string[]) {
        const changeContent = this.parseContent()
        return commentCharacters
            .map((char) => changeContent.indexOf(char) === 0)
            .some(Boolean)
    }

    format(type: 'add' | 'del'): parse.AddChange | parse.DeleteChange {
        const baseChange = {
            type,
            ln: this.ln,
            content: this.content
        }
        return type === 'add'
            ? { ...baseChange, type, add: true }
            : { ...baseChange, type, del: true }

    }
}

export abstract class BaseParser {
    abstract identity: string
    // whether to add the client name (eg. dvcClient) at the beginning of the pattern
    matchClientName = true

    // list of custom client names to also match against
    clientNames: string[]

    // regex pattern to use to match the start of the "variable" method call, eg. `.variable(`
    abstract variableMethodPattern: RegExp

    // a list of patterns to use to match each positional parameter expected in the variable call,
    // and the closing bracket, eg. user, variable, default)
    orderedParameterPatterns: RegExp[] | null = null

    // delimiter to use when detecting "named" parameters, eg. dvcClient.variable(variable=variable, default=default)
    namedParameterDelimiter = '='

    // a map of named parameter names to the patterns used to detect them eg. {user: /someregex/}
    namedParameterPatternMap: Record<string, RegExp> | null = null

    // comment characters to look for when matching. Varies by language.
    commentCharacters: string[] = ['//']

    constructor(extension: string, protected options: ParseOptions) {
        this.clientNames = [...(options.clientNames || []), 'dvcClient']
    }

    private buildParameterPattern(
        namedParameterPatternMap: Record<string, RegExp> | null,
        orderedParameterPatterns: RegExp[] | null
    ) {
        const namedParameters = namedParameterPatternMap
            ? Object
                .entries(namedParameterPatternMap)
                .map(([key, pattern]) =>
                    new RegExp(
                        `(?=.*?${key}${this.namedParameterDelimiter}\\s*${pattern.source})`
                    ).source
                )
                .join('')
                .concat(/[^)]*\)/.source)
            : null

        const orderedParameters = orderedParameterPatterns
            ? orderedParameterPatterns
                .map((p) => p.source)
                .join(/\s*,\s*/.source)
                .concat(/\)/.source)
            : null

        return orderedParameters && namedParameters
            ? new RegExp(`(?:(?:${orderedParameters})|(?:${namedParameters}))`).source
            : orderedParameters || namedParameters
    }

    private buildClientNamePattern() {
        return this.matchClientName
            ? new RegExp(`(?:${this.clientNames.join('|')})`).source
            : ''
    }

    private buildRegexPattern() {
        const clientNamePattern = this.buildClientNamePattern()

        const parameterPattern = this.buildParameterPattern(
            this.namedParameterPatternMap,
            this.orderedParameterPatterns
        )

        return new RegExp(
            clientNamePattern
            + this.variableMethodPattern.source
            + parameterPattern
        )
    }

    private extractUsageInformation(file: usage.File, match: MatchWithRange): MultilineUsageChunk {
        const lines: usage.LineItem[] = []
        let index = 0
    
        for (const line of file.lines) {
            const trimmedContent = line.content.trim()
            if (!trimmedContent.length) continue

            const { start: matchStartIndex, end: matchEndIndex } = match.range
            const lineStartIndex = index
            const lineEndIndex = lineStartIndex + trimmedContent.length - 1
            if (
                (matchStartIndex >= lineStartIndex && matchStartIndex <= lineEndIndex)
                || (lineStartIndex >= matchStartIndex && lineStartIndex <= matchEndIndex)
            ) {
                lines.push(line)
            }
            index = lineEndIndex + 1
        }

        return {
            content: lines.reduce((prev, curr) => prev + (prev ? '\n' : '') + curr.content, ''),
            start: lines.length ? lines[0].ln + 1 : -1,
            end: lines.length ? lines[lines.length - 1].ln + 1 : -1
        }
    }

    /**
     * Given a chunk from parse-diff, aggregate added and removed changes.
     * Each resulting chunk includes the multi-line content as a single string,
     * and an array of changes, each with a range object describing the start/end
     * indicies of the substring within the content string.
     */
    private aggregateMultilineChunks(chunk: parse.Chunk): { added: MultilineDiffChunk, removed: MultilineDiffChunk } {
        const added: MultilineDiffChunk = {
            content: '',
            changes: []
        }
        const removed: MultilineDiffChunk = {
            content: '',
            changes: []
        }

        const pushChangeToMultilineDiffChunk = (chunk: MultilineDiffChunk, change: ParsedChange) => {
            change.range.start = chunk.content.length

            const changeContent = change.parseContent()
            if (!changeContent.length) return

            chunk.content += changeContent
            change.range.end = chunk.content.length - 1

            chunk.changes.push(change)
        }

        for (const change of chunk.changes) {
            const parsedChange = new ParsedChange(change)
            if (parsedChange.isComment(this.commentCharacters)) continue

            if (parsedChange.type === 'add') {
                pushChangeToMultilineDiffChunk(added, parsedChange)
            } else if (parsedChange.type === 'del') {
                pushChangeToMultilineDiffChunk(removed, parsedChange)
            } else if (parsedChange.type === 'normal') {
                pushChangeToMultilineDiffChunk(added, parsedChange)
                pushChangeToMultilineDiffChunk(removed, parsedChange)
            }
        }
        return { added, removed }
    }

    printRegexPattern(): void {
        console.log(`Pattern for ${this.identity} parser: ${this.buildRegexPattern().source}`)
    }

    match(content: string): MatchResult | null {
        const pattern = this.buildRegexPattern()

        const matches = pattern.exec(content)
        if (matches) {
            const varName = (matches[1] || matches[2]).trim()
            return {
                isUnknown: !varName.match(/^["'].*["']$/),
                // position or named parameter match
                name: varName.replace(/^["']|["']$/g, ''),
                content: matches[0],
                index: matches.index
            }
        }

        return null
    }

    /**
     * Given a string, returns all matches found in the string.
     * Each match includes the variable name and the character range of the
     * matching substring.
     */
    private getAllMatches(matchString: string): MatchWithRange[] {
        const allMatches: MatchWithRange[] = []
        let cursorPosition = 0

        while (cursorPosition < matchString.length) {
            const substring = matchString.slice(cursorPosition)
            const match = this.match(substring)

            if (!match) break

            const { content, index, name, isUnknown } = match

            const startIndex = cursorPosition + index
            const endIndex = startIndex + content.length - 1

            allMatches.push({
                name,
                isUnknown,
                range: {
                    start: startIndex,
                    end: endIndex
                }
            })
            cursorPosition = endIndex + 1
        }

        return allMatches
    }

    private formatMatch(match: MatchWithRange, file: parse.File, change: parse.Change): VariableDiffMatch {
        return {
            name: match.name,
            fileName: file.to ?? '',
            line: change.type === 'normal' ? change.ln1 : change.ln,
            mode: change.type === 'add' ? 'add' : 'remove',
            ...(match.isUnknown ? { isUnknown: true } : {})
        }
    }

    /**
     * Given an array of matches, find the first change object corresponding to the match.
     * Also verify that the match is associated with at least one add/delete change object.
     */
    private formatMatches(file: parse.File, matches: MatchWithRange[], { changes }: MultilineDiffChunk) {
        const results: VariableDiffMatch[] = []

        matches.forEach((match) => {
            const { start: matchStartIndex, end: matchEndIndex } = match.range

            // Get all changes within the range of the match
            const matchingChanges = changes.filter(({ range }) => {
                const { start: changeStartIndex, end: changeEndIndex } = range
                return (matchStartIndex >= changeStartIndex && matchStartIndex <= changeEndIndex)
                    || (changeStartIndex >= matchStartIndex && changeStartIndex <= matchEndIndex)
            })

            // Remove "normal" change objects
            const validChange = matchingChanges.find(({ type }) => type !== 'normal')

            // If we have an add/del change, update the starting change to have that type and push to results
            if (validChange) {
                const formattedChange = matchingChanges[0].format(
                    validChange.type as 'add' | 'del'
                )
                results.push(this.formatMatch(match, file, formattedChange))
            }
        })

        return results
    }

    parseDiffs(file: parse.File): VariableDiffMatch[] {
        const results: VariableDiffMatch[] = []

        for (const chunk of file.chunks) {

            const { added: addedChunk, removed: removedChunk } = this.aggregateMultilineChunks(chunk)
            const addedMatches = this.getAllMatches(addedChunk.content)
            const removedMatches = this.getAllMatches(removedChunk.content)

            results.push(
                ...this.formatMatches(file, addedMatches, addedChunk),
                ...this.formatMatches(file, removedMatches, removedChunk)
            )
        }
        return results
    }

    parseFile(file: usage.File): VariableUsageMatch[] {
        const result: VariableUsageMatch[] = []
        let fileContent = ''
        for (const line of file.lines) {
            fileContent = fileContent.concat(line.content.trim()) 
        }

        const matches = this.getAllMatches(fileContent)

        for (const match of matches) {
            const usage = this.extractUsageInformation(file, match)
            if (usage.start !== -1 && usage.end !== -1) {
                result.push({
                    name: match.name,
                    line: usage.start,
                    fileName: file.name,
                    content: usage.content
                })
            }
        }

        return result
    }
}
