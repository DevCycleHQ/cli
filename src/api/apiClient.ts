import axios, { AxiosError } from 'axios'
import { BASE_URL } from './common'
import { createApiClient } from './zodClient'
import { ZodIssueCode, ZodIssueOptionalMessage, ErrorMapCtx } from 'zod'

export const axiosClient = axios.create({
    baseURL: BASE_URL,
})

export const setDVCReferrer = (
    command = 'unknown',
    version: string,
    caller = 'cli',
): void => {
    axiosClient.defaults.headers.common['dvc-referrer'] = 'cli'
    axiosClient.defaults.headers.common['dvc-referrer-metadata'] = JSON.stringify({
        command,
        version,
        caller,
    })
}

axiosClient.interceptors.response.use(
    (response) => {
        return response
    },
    (error: AxiosError) => {
        if (error.response?.status === 401) {
            console.info('Authorization Error: Please login using "dvc login again".')
        } else {
            const responseData = error.response?.data as Record<string, any>
            console.info('DevCycle Error:', responseData?.message)
        }
        // TODO: Handle this error properly, DVC-7758
        return Promise.reject(error)
    },
)

export const errorMap = (issue: ZodIssueOptionalMessage, ctx: ErrorMapCtx) => {
    if (!ctx.data) {
        return {
            message: `${issue.path.join('.')} is a required field`,
        }
    }

    switch (issue.code) {
        case ZodIssueCode.too_small:
            return {
                message:
                issue.type === 'string' ?
                    `${issue.path.join('.')} length must be >= ${issue.minimum} but got ${ctx.data}`
                    : `${issue.path.join('.')} must be >= ${issue.minimum} but got ${ctx.data}`,
            }
        case ZodIssueCode.too_big:
            return {
                message: 
                issue.type === 'string' ?
                    `${issue.path.join('.')} length must be <= ${issue.maximum} but got ${ctx.data}` 
                    : `${issue.path.join('.')} must be <= ${issue.maximum} but got ${ctx.data}`,
            }
        case ZodIssueCode.invalid_string:
            const regexError = `format of ${issue.path.join('.')} is invalid.`
            const invalidTypeError = `${issue.path.join('.')} must be a valid ${issue.validation}, but got ${ctx.data}`
            const noValidationError = `Invalid value: ${issue.path.join('.')}. ${issue.message || ''}`
            
            return {
                message: issue.validation === 'regex' ? regexError 
                    : issue.validation ? 
                        invalidTypeError 
                        : noValidationError,
            }
        case ZodIssueCode.invalid_enum_value:
            return {
                message: `${issue.path.join('.')} must be one of ${issue.options.join(', ')}`,
            }
        default:
            return { message: ctx.defaultError }
    }
}

export const apiClient = createApiClient(BASE_URL, { axiosInstance: axiosClient, validate: 'request' })
export default apiClient
