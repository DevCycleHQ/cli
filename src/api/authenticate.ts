import axios from 'axios'
import { AUTH_URL } from './common'

export const authenticate = async (client_id: string, client_secret: string): Promise<string> => {
    const url = new URL("/oauth/token", AUTH_URL)

    const response = await axios.post(url.href, {
        grant_type: "client_credentials",
        client_id,
        client_secret,
        audience: "https://api.devcycle.com",
    })

    return response.data.access_token
}
