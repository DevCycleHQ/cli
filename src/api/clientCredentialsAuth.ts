import axios from 'axios'
import { AUTH_URL } from './common'

export const clientCredentialsAuth = async (client_id: string, client_secret: string): Promise<string> => {
    const url = new URL('/oauth/token', AUTH_URL)

    try {
        const response = await axios.post(url.href, {
            grant_type: 'client_credentials',
            client_id,
            client_secret,
            audience: 'https://api.devcycle.com/',
        })

        return response.data.access_token
    } catch (e) {
        throw new Error('Failed to authenticate with the DevCycle API. Check your credentials.')
    }

}
