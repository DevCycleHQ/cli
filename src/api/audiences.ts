import apiClient from './apiClient'
import { buildHeaders } from './common'

const BASE_URL = '/v1/projects/:project/audiences'

export const fetchAudiences = async (
    token: string,
    project_id: string
) => {
    return apiClient.get(`${BASE_URL}`, {
        headers: buildHeaders(token),
        params: {
            project: project_id,
        }
    })
}