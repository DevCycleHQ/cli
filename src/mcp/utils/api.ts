import { DevCycleAuth } from './auth'
import { fetchFeatures, createFeature } from '../../api/features'
import { fetchVariables } from '../../api/variables'
import { fetchEnvironmentByKey } from '../../api/environments'
import { enableTargeting, disableTargeting } from '../../api/targeting'

function getErrorMessage(error: unknown): string {
    if (error instanceof Error) {
        return error.message
    }
    return String(error)
}

function ensureError(error: unknown): Error {
    if (error instanceof Error) {
        return error
    }
    return new Error(String(error))
}

export class DevCycleApiClient {
    constructor(private auth: DevCycleAuth) {}

    async listFeatures(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error('MCP listFeatures params:', JSON.stringify(args, null, 2))

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const authToken = this.auth.getAuthToken()
            const projectKey = this.auth.getProjectKey()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchFeatures(authToken, projectKey, query)
            console.error(
                'MCP listFeatures result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listFeatures error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            console.error('Error during listFeatures:', getErrorMessage(error))
            throw ensureError(error)
        }
    }

    async listVariables(args: {
        search?: string
        page?: number
        per_page?: number
    }) {
        console.error(
            'MCP listVariables params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const query = {
                search: args.search,
                page: args.page,
                perPage: args.per_page,
            }

            const result = await fetchVariables(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                query,
            )
            console.error(
                'MCP listVariables result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP listVariables error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async getSdkKeys(
        environmentKey: string,
        keyType?: 'mobile' | 'server' | 'client',
    ) {
        console.error(
            'MCP getSdkKeys params:',
            JSON.stringify({ environmentKey, keyType }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const environment = await fetchEnvironmentByKey(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                environmentKey,
            )

            const sdkKeys = environment.sdkKeys

            let result
            if (keyType) {
                result = {
                    [keyType]: sdkKeys[keyType],
                }
            } else {
                result = {
                    mobile: sdkKeys.mobile,
                    server: sdkKeys.server,
                    client: sdkKeys.client,
                }
            }

            console.error(
                'MCP getSdkKeys result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP getSdkKeys error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async enableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP enableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await enableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP enableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP enableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async disableTargeting(featureKey: string, environmentKey: string) {
        console.error(
            'MCP disableTargeting params:',
            JSON.stringify({ featureKey, environmentKey }, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            const result = await disableTargeting(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureKey,
                environmentKey,
            )
            console.error(
                'MCP disableTargeting result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP disableTargeting error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }

    async createFeature(args: {
        key?: string
        name?: string
        description?: string
        type?: 'release' | 'experiment' | 'permission' | 'ops'
        interactive?: boolean
    }) {
        console.error(
            'MCP createFeature params:',
            JSON.stringify(args, null, 2),
        )

        try {
            this.auth.requireAuth()
            this.auth.requireProject()

            if (args.interactive) {
                // For interactive mode, we would need to implement prompts
                // For now, return an error asking for explicit parameters
                throw new Error(
                    'Interactive mode not yet supported in MCP. Please provide explicit parameters: key, name, description, type',
                )
            }

            if (!args.key || !args.name) {
                throw new Error(
                    'Feature key and name are required when not using interactive mode',
                )
            }

            const featureData = {
                key: args.key,
                name: args.name,
                description: args.description || '',
                type: args.type || 'release',
            }

            const result = await createFeature(
                this.auth.getAuthToken(),
                this.auth.getProjectKey(),
                featureData,
            )
            console.error(
                'MCP createFeature result:',
                JSON.stringify(result, null, 2),
            )
            return result
        } catch (error) {
            console.error(
                'MCP createFeature error:',
                JSON.stringify({ error: getErrorMessage(error) }, null, 2),
            )
            throw ensureError(error)
        }
    }
}
