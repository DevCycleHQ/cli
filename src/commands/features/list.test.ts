import { expect } from '@oclif/test'
import { dvcTest } from '../../../test-utils'
import { BASE_URL } from '../../api/common'

describe('features list', () => {
    const projectKey = 'test-project'
    const authFlags = ['--client-id', 'test-client-id', '--client-secret', 'test-client-secret']

    const mockFeatures = [
        {
            key: 'feature-1',
            name: 'Feature 1',
            _id: '61450f3daec96f5cf4a49946'
        },
        {
            key: 'feature-2',
            name: 'Feature 2',
            _id: '61450f3daec96f5cf4a49947',
        }
    ]

    dvcTest()
        .nock(BASE_URL, (api) => api
            .get(`/v1/projects/${projectKey}/features`)
            .reply(200, mockFeatures)
        )
        .stdout()
        .command(['features list', '--project', projectKey, ...authFlags])
        .it('returns a list of feature keys', (ctx) => {
            expect(ctx.stdout).to.contain(JSON.stringify(['feature-1', 'feature-2'], null, 2))
        })

    dvcTest()
        .nock(BASE_URL, (api) => api
            .get(`/v1/projects/${projectKey}/features`)
            .query({ page: 2, perPage: 10 })
            .reply(200, mockFeatures)
        )
        .stdout()
        .command([
            'features list',
            '--project', projectKey,
            '--page', '2',
            '--per-page', '10',
            ...authFlags
        ])
        .it('passes pagination params to api', (ctx) => {
            expect(ctx.stdout).to.contain(JSON.stringify(['feature-1', 'feature-2'], null, 2))
        })
    
    dvcTest()
        .nock(BASE_URL, (api) => api
            .get(`/v1/projects/${projectKey}/features`)
            .query({ search: 'hello world' })
            .reply(200, mockFeatures)
        )
        .stdout()
        .command([
            'features list',
            '--project', projectKey,
            '--search', 'hello world',
            ...authFlags
        ])
        .it('passes search param to api', (ctx) => {
            expect(ctx.stdout).to.contain(JSON.stringify(['feature-1', 'feature-2'], null, 2))
        })
})
