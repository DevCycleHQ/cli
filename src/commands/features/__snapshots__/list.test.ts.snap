// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`features list > passes pagination params to api 1`] = `
"[]
"
`;

exports[`features list > passes search param to api 1`] = `
"[]
"
`;

exports[`features list > returns a list of feature keys 1`] = `
"[
  "first-feature",
  "second-feature"
]
"
`;
